### ISSUE TITLE
JS Basic Data Structures: Arrays & Objects

### ISSUE THREAD

**Issue by bonham000** [FreeCodeCamp/CurriculumExpansion#90]
## This Issue Contains the Proposed Basic Data Structures Challenges

As referenced in the [proposed JavaScript section](https://github.com/FreeCodeCamp/CurriculumExpansion/issues/89) to the Curriculum Expansion, we are adding a **basic data structures** section to the JavaScript curriculum. This issue contains the development of these challenges. All other [advanced data structures](https://github.com/FreeCodeCamp/CurriculumExpansion/issues/17) will reside in the later optional section.

These challenges are assigned to @no-stack-dub-sack & @bonham000.

### Arrays:
- [x] [Use an array to store a collection of data](#issuecomment-271119327)
- [x] [Access an array's contents using bracket notation](#issuecomment-270932804)
- [x] [Add items to an array with `push()` and `unshift()`](#issuecomment-271119750)
- [x] [Remove items from an array with  `pop()` and `shift()`](#issuecomment-271120009)
- [x] [Remove items using `splice()`](#issuecomment-271120314)
- [x] [Add items using `splice()`](#issuecomment-271120709)
- [x] [Copy an array with `slice()`](#issuecomment-271121015)
- [x] [Combining arrays with spread syntax](#issuecomment-271121216)
- [x] [Insert elements with spread syntax](#issuecomment-271123392)
- [x] [Check for the presence of an item with `indexOf()`](#issuecomment-271121351)
- [x] [Iterate through all the items in an array](#issuecomment-271121660)
- [x] [Create a complex nested array](#issuecomment-271121845)

### Objects:
- [x] [Add Key-Value Pairs to JavaScript Objects](#issuecomment-270027987)
- [x] [Modify an Object Nested Within an Object](#issuecomment-270028442)
- [x] [Access Property Names with Bracket Notation](#issuecomment-270028960)
- [x] [Use the Delete Keyword to Remove Object Properties](#issuecomment-270029440)
- [x] [Check if an Object has a Property](#issuecomment-270029755)
- [x] [Iterate Through the Keys of an Object with a `for...in` Statement](#issuecomment-270030267)
- [x] [Generate an Array of All Object Keys with `Object.keys()`](#issuecomment-270030546)
- [x] [Modify an Array Stored in an Object](#issuecomment-270031008)

---

**Comment by bonham000** [comment]
<b>Title:</b> Add Key-Value Pairs to JavaScript Objects

<b>Description:</b> The next data structure we will discuss is the JavaScript `object`. Objects are a very fundamental part of JavaScript, in fact, you may have heard this line before: 'In JavaScript, everything is an object.' While an understanding of objects is important to understand the inner workings of JavaScript functions or JavaScript's object-oriented capabilities, here we will confine our discussion to the use of JavaScript objects as `key-value` pair data structures.

Key-value pair data structure go by different names depending on the language you are using and the specific details of the data structure. The terms <i>dictionary</i>, <i>map</i>, and <i>hash table</i> all refer to the notion of a data structure in which specific <i>keys</i>, or <i>properties</i> are mapped to specific <i>values</i>. For instance, consider the following:

``` javascript
let FCC_User = {
 username: 'awesome_coder',
 followers: 572,
 points: 1741,
 completedProjects: 15
};
```

We've just defined an object called `FCC_User` with four <i>properties</i> each of which map to a specific <i>value</i>. If you wanted to know the number of followers `FCC_User` has, you could access that property by writing `FCC_User.followers` — this is called dot notation. You could also access the property with brackets, like so: `FCC_User['followers']`. Notice with the bracket notation we enclosed  `followers` in quotes. This is because the brackets actually allow us to pass a variable in to be evaluated as a property name. Keep this in mind for later.

<b>Instructions:</b> We've created a `foods` object here with three entries. Add three more entries: `bananas` with a value of `13`, `grapes` with a value of `35`, and `strawberries` with a value of `27`.

``` javascript
// SEED CODE
let foods = {
	apples: 25,
	oranges: 32,
	plums: 28
};
// change code below this line
```
``` javascript
// SOLUTION CODE
let foods = {
	apples: 25,
	oranges: 32,
	plums: 28
};

foods.bananas = 13;
foods.grapes = 35;
foods.strawberries = 27;
```
``` javascript
// TESTS
assert(typeof foods === 'object', 'foods is an object');
assert(foods.bananas === 13, 'The foods object has a key \'bananas\' with a value of 13.');
assert(foods.grapes === 35, 'The foods object has a key \'grapes\' with a value of 35.');
assert(foods.strawberries === 27, 'The foods object has a key \'strawberries\' with a value of 27.');
```

**Comment by bonham000** [comment]
<b>Title:</b> Modify an Object Nested Within an Object

<b>Description:</b> Objects, and other similar key-value pair data structures, offer some very useful benefits. One clear benefit is that they allow us to structure our data in an intuitive way. They are also very flexible. For instance, you can have properties nested to an arbitrary depth. Values can also be anything, for example a key can store an array, or even another object. Objects are also the foundation for JavaScript Object Notation, JSON, which is a widely used method of sending data across the web.

Another powerful advantage of key-value pair data structures is constant lookup time. What we mean by this is when you request the value of a specific property you will get the value back in the same amount of time (theoretically) regardless of the number of entries in the object. If you had an object with 5 entries or one that held a collection of 1,000,000 entries you could still retrieve property values or check if a key exists in the same amount of time.

The reason for this fast lookup time is that internally the object is storing properties using some type of <i>hashing mechanism</i> which allows it to know exactly where it has stored different property values. If you want to learn more about this please take a look at the optional Advanced Data Structures challenges. All you should remember for now is that performant access to flexibly structured data make key-value stores very attractive data structures useful in a wide variety of settings.

<b>Instructions:</b> Here we've written an object `nestedObject` which includes another object nested within it. You can modify properties on this nested object in the same way you modified properties in the last challenge. Set the value of the `online` key to `45`.

``` javascript
// SEED CODE
let nestedObject = {
	id: 23894201352,
	date: 'January 1, 2017',
	data: {
		totalUsers: 51,
		online: 42
	}
};
// change code below this line
```
``` javascript
// SOLUTION CODE
let nestedObject = {
	id: 23894201352,
	date: 'January 1, 2017',
	data: {
		totalUsers: 51,
		online: 42
	}
};

nestedObject.data.online = 45;
```
``` javascript
// TESTS
assert('id' in nestedObject && 'date' in nestedObject && 'data' in nestedObject, 'nestedObject has id, date and data properties.');
assert('totalUsers' in nestedObject.data && 'online' in nestedObject.data, 'nestedObject has a data key set to an object with keys totalUsers and online.');
assert(nestedObject.data.online === 45, 'The online property nested in the data key of nestedObject should be set to 45.');
```

**Comment by bonham000** [comment]
<b>Title:</b> Access Property Names with Bracket Notation

<b>Description:</b> In the first challenge we mentioned the use of bracket notation as a way access properties values using the evaluation of a variable. For instance, if you recall our `foods` object from that challenge, imagine that this object is being used in a program for a supermarket cash register. We have some function that sets the `selectedFood` and we want to check our `foods` object for the presence of that food. This might look like:<br>

``` javascript
let selectedFood = getCurrentFood(scannedItem);
let inventory = foods[selectedFood];
```

This code will evaluate the value stored in the `selectedFood` variable and return the value of that key in the `foods` object, or `undefined` if it is not present. Bracket notation is very useful because sometime object properties are not known before runtime or we need to access them in a more dynamic way.

<b>Instructions:</b> In the example code we've defined a function `checkInventory` which receives a scanned item as an argument. Return the current value of the `scannedItem` key in the `foods` object. You can assume that only valid keys will be provided as an argument to `checkInventory`.

``` javascript
// SEED CODE
let foods = {
	apples: 25,
	oranges: 32,
	plums: 28,
	bananas: 13,
	grapes: 35,
	strawberries: 27
};

function checkInventory(scannedItem) {
	// change code below this line

	// change code above this line
};
```
``` javascript
// SOLUTION CODE
let foods = {
	apples: 25,
	oranges: 32,
	plums: 28,
	bananas: 13,
	grapes: 35,
	strawberries: 27
};

function checkInventory(scannedItem) {
	return foods[scannedItem];
};
```
``` javascript
// TESTS
assert(typeof checkInventory === 'function', 'checkInventory is a function');
assert('apples' in foods && 'oranges' in foods && 'plums' in foods && 'bananas' in foods && 'grapes' in foods && 'strawberries' in foods, 'The foods object contains the following keys: apples, oranges, plums, bananas, grapes, and strawberries.');
assert(checkInventory('apples') === 25 && checkInventory('bananas') === 13 && checkInventory('strawberries') === 27, 'The checkInventory function returns the value of the scannedItem argument in the foods object.');
```

**Comment by bonham000** [comment]
<b>Title:</b> Use the Delete Keyword to Remove Object Properties

<b>Description:</b> Now you know what objects are and their basic features and advantages. In short, they are key-value stores which provide a flexible, intuitive way to structure data and they provide very fast lookup time. For the rest of these challenges, we will describe several common operations you can perform on objects so you can become comfortable applying these useful data structures in your programs.

Previously, we added and modified key-value pairs to objects. Here we will see how we can remove a key-value pair from an object. If we wanted to remove the `apples` key from our `foods` object from before, we could remove it by using the `delete` keyword like this:

``` javascript
delete foods.apples;
```

<b>Instructions:</b> Use the `delete` keyword to remove the `oranges`, `plums`, and `strawberries` keys from the `foods` object.
``` javascript
// SEED CODE
let foods = {
	apples: 25,
	oranges: 32,
	plums: 28,
	bananas: 13,
	grapes: 35,
	strawberries: 27
};
// change code below this line
```
``` javascript
// SOLUTION CODE
let foods = {
	apples: 25,
	oranges: 32,
	plums: 28,
	bananas: 13,
	grapes: 35,
	strawberries: 27
};

delete foods.oranges;
delete foods.plums;
delete foods.bananas;
```
``` javascript
// TESTS
assert(!foods.hasOwnProperty('oranges') && !foods.hasOwnProperty('plums') && !foods.hasOwnProperty('bananas') && Object.keys(foods).length === 3, 'The foods object only has three keys: apples, grapes, and strawberries.');
```

**Comment by bonham000** [comment]
<b>Title:</b> Check if an Object has a Property

<b>Description:</b> Now we can add, modify, and remove keys from objects. But what if we just wanted to know if an object has a specific property? JavaScript provides two different ways to do this. One uses the `hasOwnProperty` method on the object and the other uses the `in` keyword. If we have an object `users` with a property of `Alan`, we could check for its presence in either of the following ways:

``` javascript
users.hasOwnProperty('Alan');
'Alan' in users;
// both return true
```

<b>Instructions:</b> We've created a `users` object here with some users in it and a function `isEveryoneHere` which we pass the `users` object to as an argument. Finish writing this function so that it returns `true` only if the `users` object contains all four names, Alan, Jeff, Sarah, and Ryan, as keys, and `false` otherwise.

``` javascript
SEED CODE
let users = {
	Alan: {
		age: 27,
		online: true
	},
	Jeff: {
		age: 32,
		online: true
	},
	Sarah: {
		age: 48,
		online: true
	},
	Ryan: {
		age: 19,
		online: true
	}
};

function isEveryoneHere(obj) {
	// change code below this line

	// change code above this line
};

isEveryoneHere(users);
```
``` javascript
// SOLUTION CODE
let users = {
	Alan: {
		age: 27,
		online: true
	},
	Jeff: {
		age: 32,
		online: true
	},
	Sarah: {
		age: 48,
		online: true
	},
	Ryan: {
		age: 19,
		online: true
	}
};

function isEveryoneHere(obj) { 
	return ('Alan' in users && 'Jeff' in users && 'Sarah' in users && 'Ryan' in users) ? true : false;
};

isEveryoneHere(users);
```
``` javascript
// TESTS
assert('Alan' in users && 'Jeff' in users && 'Sarah' in users && 'Ryan' in users && Object.keys(users).length === 4, 'The users object only contains the keys Alan, Jeff, Sarah, and Ryan.');
assert(isEveryoneHere(users) === true, 'The function isEveryoneHere returns true if Alan, Jeff, Sarah, and Ryan are properties on the users object');
assert((function() { delete users.Alan; delete users.Jeff; delete users.Sarah; delete users.Ryan; return isEveryoneHere(users) })() === false, 'The function isEveryoneHere returns false if Alan, Jeff, Sarah, and Ryan are not keys on the users object.');
```

**Comment by bonham000** [comment]
<b>Title:</b> Iterate Through the Keys of an Object with a for...in Statement

<b>Description:</b> Sometimes you may need to iterate through all the keys within an object. This requires a specific syntax in JavaScript called a for...in statement. For our `users` object from the last challenge, this could look like:

``` javascript
for (let user in users) {
 console.log(user);
};
```

In this statement, we define a variable `user`. This variable will be set to the key in each iteration as the statement loops through the keys in the object. Running this code would print the name of each user to the console. Note that objects do not maintain an ordering to stored keys like arrays do.

<b>Instructions:</b> We've defined a function `countOnline` that should return the number of users with the online property set to `true`. Use a `for...in` statement within this function to loop through the users in the users object and return the number of users whose online property is set to `true`.

``` javascript
// SEED CODE
let users = {
	Alan: {
		age: 27,
		online: false
	},
	Jeff: {
		age: 32,
		online: true
	},
	Sarah: {
		age: 48,
		online: false
	},
	Ryan: {
		age: 19,
		online: true
	}
};

function countOnline(obj) {
	// change code below this line

	// change code above this line
};

countOnline(users);
```
``` javascript
// SOLUTION CODE
let users = {
	Alan: {
		age: 27,
		online: false
	},
	Jeff: {
		age: 32,
		online: true
	},
	Sarah: {
		age: 48,
		online: false
	},
	Ryan: {
		age: 19,
		online: true
	}
};

function countOnline(obj) { 
	let n = 0;
	for (let user in obj) if (obj[user].online) n++;
	return n;
};

countOnline(users);
```
``` javascript
// TESTS
assert(users.Alan.online === false && users.Jeff.online === true &&  users.Sarah.online === false &&  users.Ryan.online === true, 'The users object contains users Jeff and Ryan with online set to true and users Alan and Sarah with online set to false.');
assert((function() { users.Harry = {online: true}; users.Sam = {online: true}; users.Carl = {online: true}; return countOnline(users) })() === 5, 'The function countOnline returns the number of users with the online property set to true.');
```

**Comment by bonham000** [comment]
<b>Title:</b> Generate an Array of All Object Keys with Object.keys()

<b>Description:</b> We can also generate an array which contains all the keys stored in an object using the `Object.keys()` method and passing in an object as the argument. This will return an array with strings representing each property in the object. Again, there will be no specific order to the entries in the array.

<b>Instructions:</b> Finish writing the `getArrayOfUsers` function so that it returns an array containing all the properties in the object it receives as an argument.
``` javascript
// SEED CODE
let users = {
	Alan: {
		age: 27,
		online: false
	},
	Jeff: {
		age: 32,
		online: true
	},
	Sarah: {
		age: 48,
		online: false
	},
	Ryan: {
		age: 19,
		online: true
	}
};

function getArrayOfUsers(obj) {
	// change code below this line

	// change code above this line
};
```
``` javascript
// SOLUTION CODE
let users = {
	Alan: {
		age: 27,
		online: false
	},
	Jeff: {
		age: 32,
		online: true
	},
	Sarah: {
		age: 48,
		online: false
	},
	Ryan: {
		age: 19,
		online: true
	}
};

function getArrayOfUsers(obj) {
	return Object.keys(obj);
};
```
``` javascript
// TESTS
assert('Alan' in users && 'Jeff' in users && 'Sarah' in users && 'Ryan' in users && Object.keys(users).length === 4, 'The users object only contains the keys Alan, Jeff, Sarah, and Ryan.');
assert((function() { users.Sam = {}; users.Lewis = {}; let R = getArrayOfUsers(users); return (R.indexOf('Alan') !== -1 && R.indexOf('Jeff') !== -1 && R.indexOf('Sarah') !== -1 && R.indexOf('Ryan') !== -1 && R.indexOf('Sam') !== -1 && R.indexOf('Lewis') !== -1); })() === true, 'The getArrayOfUsers function returns an array which contains all the keys in the users array.');
```

**Comment by bonham000** [comment]
<b>Title:</b> Modify an Array Stored in an Object

<b>Description:</b> Now you've seen all the basic operations for JavaScript objects. You can add, modify, and remove key-value pairs, check if keys exist, and iterate over all the keys in an object. As you continue learning JavaScript you will see even more versatile applications of objects. Additionally, the optional Advanced Data Structures lessons later in the curriculum also cover the ES6 `Map` and `Set` objects, both of which are similar to ordinary objects but provide some additional features. Now that you've learned the basic of arrays and objects, you're ready to put this knowledge to use solving some algorithms!

<b>Instructions:</b> Take a look at the object we've provided in the code editor. The `user` object contins three keys. The `data` key contains four keys, one of which contains an array of friends. From this you can see how flexible objects are as data structures. We've started writing a function `addFriend`. Finish writing it so that it takes a `user` object and adds the name of the `friend` argument to the array stored in `user.data.friends`.

``` javascript
// SEED CODE
let user = {
	name: 'Kenneth',
	age: 28,
	data: {
		username: 'kennethCodesAllDay',
		joinDate: 'March 26, 2016',
		organization: 'Free Code Camp',
		friends: [
			'Sam',
			'Kira',
			'Tomo'
		],
		location: {
			city: 'San Francisco',
			state: 'CA',
			country: 'USA'
		}
	}
};

function addFriend(user, friend) {
	// change code below this line	

	// change code above this line
};
```
``` javascript
// SOLUTION CODE
let user = {
	name: 'Kenneth',
	age: 28,
	data: {
		username: 'kennethCodesAllDay',
		joinDate: 'March 26, 2016',
		organization: 'Free Code Camp',
		friends: [
			'Sam',
			'Kira',
			'Tomo'
		],
		location: {
			city: 'San Francisco',
			state: 'CA',
			country: 'USA'
		}
	}
};

function addFriend(user, friend) {
	user.data.friends.push(friend);
	return user;
};
```
``` javascript
// TESTS
assert('name' in user && 'age' in user && 'data' in user, 'The user object has name, age, and data keys');
assert((function() { let L1 = user.data.friends.length; addFriend(user, 'Sean'); let L2 = user.data.friends.length; return (L2 === L1 + 1); })(), 'The addFriend function accepts a user object and a friend as arguments and adds the friend to the array of friends in the user object.');
```

**Comment by QuincyLarson** [comment]
@bonham000 nice work. These look solid. Excellent explanations and examples so far.

**Comment by no-stack-dub-sack** [comment]
@QuincyLarson @bonham000 Should have these done by friday evening

**Comment by no-stack-dub-sack** [comment]
@bonham000 @QuincyLarson - I know we discussed trying not to include MDN links, but I think it is important for challenges like these. Mainly because, with topics like `splice()`, we simply can't cover everything without being overly verbose - for instance, covering the things like negative values for parameters and the result of omitting the second parameter - covering these intricacies would likely make the challenge convoluted and too wordy, but these are important things to learn. What do you guys think? 

If we don't include the external links, we run the risk of it being said that the challenges aren't thorough or don't provide deep enough scope on a given topic. Including them covers our bases without us having to explicitly cover it.

**Comment by smileytechguy** [comment]
I like the MDN links personally, very helpful for some api doc

**Comment by smileytechguy** [comment]
are we going to use let as well now?  We will need to cover that in the initial part of javascript if you havent already

**Comment by no-stack-dub-sack** [comment]
@smileytechguy For sure, most of the new ES6 features are going to be covered in the curriculum expansion - you can see a list of topics for that section here #34.

**Comment by QuincyLarson** [comment]
@no-stack-dub-sack We will want to tell campers that "roughly, splice does..." and give them a high level overview. We aren't aiming to provide comprehensive definitions and descriptions. Rather, we just want them to understand it well enough to solve the challenge and keep moving. 

The MDN is good, but it can be intimidating, and sometimes it takes several re-reads to understand the concept they're talking about. I say do your best to explain these briefly, and our community can amend your description as necessary in the future.

**Comment by smileytechguy** [comment]
@no-stack-dub-sack what about older browsers?  They won't work with FCC now...

**Comment by smileytechguy** [comment]
# Use an array to store a collection of data

## Description

We have learned that we can use the `let` and `var` keywords in order to store data.  However, what if we want to store a collection of data?

Imagine you had a bunch of coffee cups, and wanted to put them on a shelf.  Rather than building an individual shelf for each cup, you would put them all on the same shelf.

In code, this also makes logical sense.  Rather than making a variable for each one such as `coffeCup1`, `coffeCup2`, `coffeCup3`, `coffeCup4`, `coffeCup5`, you would make a "shelf" of them.

In JavaScript, these "shelfs" are called *arrays*.  An array can store multiple variables in one variable, keeping them in order and much more.

In order to create an array, you use the following syntax:
```js
var myArray = [];
```
**saving so i dont lose this draft, will edit and resave when done**

**Comment by smileytechguy** [comment]
# Accessing an array's contents

I would like to add this challenge, as it is too much to combine with the above.

## Description

In the previous challenge, we learned how to create an array.  Now, we want to access this information.

When we define a sample array as seen below, there are 3 items in it:
```js
var ourArray = ["a", "b", "c"];
```

In an array, each array item has an index.  This index doubles as the position in the array, as well as how you reference it.

However, compared to real life, these arrays start at index `0`.

In order to get something from an array, we can use the index in combination with brackets.

For example, if we want to get the `"a"` from `ourArray` and assign it to a variable, we can do so with the following snippet:

```js
var ourVariable = ourArray[0];
```

Besides accessing the value associated with an index, you can set an index to a value:

```js
ourArray[1] = "not b anymore";
```

The above line would set index 1 in `ourArray`, which was `"b"`, to `"not b anymore"`.

In order to complete this challenge, set the 2^nd position (1^st index)

## Seed code

```js
// create your array here!
```

## Solution code

```js
// create your array here!
var myArray = ["a", "b", "c"];
```
or
```js
// create your array here!
let myArray = ["a", "b", "c"];
```

## Tests

```js
assert(typeof myArray === 'object', 'myArray is an array');
assert(myArray === ["a", "b", "c"], 'myArray contains `"a"`, `"b"`, and `"c"`');
```

**Comment by smileytechguy** [comment]
I will try to do all of the following:
```
 Remove items from an array with pop() and shift()
 Remove items using splice()
 Add items using splice()
 Copy an array with slice()
 Copy an array with spread syntax
 Check for the presence of an item with indexOf()
 Iterate through all the items in an array
 Create a nested array
```

**Comment by bonham000** [comment]
Hey @smileytechguy I believe at the moment @no-stack-dub-sack is working on developing these array challenges, I think he already has some work completed for several of these, just not finalized and added here yet.

**Comment by smileytechguy** [comment]
@bonham000 oh alright, sure

**Comment by bonham000** [comment]
However, @smileytechguy, while we wait for his response if you are still looking to contribute here, the challenges on `objects` that I've already added have not been reviewed or QA'd by anyone. This will still need to occur before these challenges are actually seeded into the curriculum. If you're interested in looking over them to review them, that would be really helpful.

**Comment by smileytechguy** [comment]
@bonham000 link for the lazy?

**Comment by smileytechguy** [comment]
@bonham000 ill review them in a few

**Comment by no-stack-dub-sack** [comment]
@bonham000 @smileytechguy Hey guys, sorry for the confusion - I suppose I should have been at least checking off what was done as I completed them locally. I was just waiting til they were all done to post them to make it more streamlined. I have just checked off the ones I have completed and should be getting to the rest tonight. 

@smileytechguy 
> what about older browsers? They won't work with FCC now... 

Is this from our conversation about ES6?

**Comment by QuincyLarson** [comment]
@no-stack-dub-sack Ok - excellent. I'm stoked to see them and get them imported into our curriculum.

**Comment by smileytechguy** [comment]
@no-stack-dub-sack yes, that was about the conversion to ES6 . if there are any you want me to do, let me know

**Comment by no-stack-dub-sack** [comment]
@smileytechguy Sure! I've gotten all done but one at this point, the `insert items with spread syntax` challenge, which I just added to the list as sort of an afterthought, but if you want to do it, go for it! 

The idea we had in having @bonham000 and I each tackle one of these sections in whole, was to keep the narrative consistent throughout, so if you'd like to do it, if you don't mind, just take a quick look at how I did the preceding spread syntax challenge [here](https://github.com/no-stack-dub-sack/array-challenges/blob/master/challenge-files/challenge_7.js), and try to keep the narrative/writing consistent so our users will have a clean experience moving from one challenge to the next. 

I think the challenge you wrote on bracket notation was much needed, and I've added it to the list, but if it's cool with you, I might make some small edits before we seed it so that it matches the rest of the ones I wrote in order to keep with the same theme of keeping narrative clear and consistent. 

Thoughts? Thanks again for pitching in!

**Comment by no-stack-dub-sack** [comment]
@bonham000 @QuincyLarson Though I've checked them as complete, I have not posted them yet. Will be doing that early tomorrow once I give them a last once over to hopefully minimize the work on QAs end.

**Comment by no-stack-dub-sack** [comment]
@smileytechguy As mentioned in my previous post, I went through and made a couple of changes to your "Access elements" challenge to keep the narrative clean throughout. I'm also going to post my version of the first challenge because it sets up some other things for the rest of the challenges. Sorry for the confusion, I just hadn't posted mine yet since the OP mentioned they had been assigned.

**Comment by no-stack-dub-sack** [comment]
# Use an array to store a collection of data

## Description:
Arrays are JavaScript's most fundamental, and perhaps most common, data structure. An <strong>array</strong> is simply a collection of data, of any length (zero-indexed), arranged in a comma separated list and enclosed in brackets <code>[ ]</code>. While we often make the distinction in JavaScript between Objects and Arrays, it is important to note that technically, an <strong>array</strong> <em>is</em> a type of <strong>object</strong>.

Arrays can store any type of data supported by JavaScript, and while they are a simple and basic form of data structure, they can also be very complex and powerful - all of which depends on how the programmer utilizes them.

The below is an example of a valid array, notice it contains booleans, strings, numbers, other arrays (this is called a nested, or multi-dimensional array), and objects, among other valid data types: 

```js
let myArray = [undefined, null, true, false, 'one', 2, "III", {'four': 5}, [6, 'seven', 8]];
```

JavaScript offers many built in <em>methods</em> which allow us to access, traverse, and mutate arrays as needed, depending on our purpose. In the coming challenges, we will discuss several of the most common and useful methods, and a few other key techniques, that will help us to better understand and utilize arrays as data structures in JavaScript.

## Instructions
We have defined a variable called <code>yourArray</code>; complete the declaration by defining an array of at least 5 elements in length. Your array should contain at least one string, one number, and one boolean.

## Challenge Seed:
```js
let yourArray = // change code here;
```

## Challenge Solution:
```js
let yourArray = [true, false, 'one', 2, 'III'];
```

## Tests:
```js
assert.strictEqual(Array.isArray(yourArray), true, 'yourArray is an array.');
assert(yourArray.length >= 5, 'yourArray is at least 5 elements long.')
assert(yourArray.filter( el => typeof el === 'boolean').length >= 1, '<code>yourArray</code> contains at least one boolean.');
assert(yourArray.filter( el => typeof el === 'number').length >= 1, '<code>yourArray</code> contains at least one number.');
assert(yourArray.filter( el => typeof el === 'string').length >= 1, '<code>yourArray</code> contains at least one string.');
```

**Comment by no-stack-dub-sack** [comment]
# Add items to an array with <code>push()</code> and <code>unshift()</code>

## Description:
An array's length, like the data types it can contain, is not fixed. Arrays can be defined with a length of any number of elements, and elements can be added or removed over time. In this challenge, we will look at two methods with which we can programmatically modify an array: <code>Array.push()</code> and <code>Array.unshift()</code>. <br><br>

Both methods take one or more elements as <em>parameters</em> and add those elements to the array the method is being called on; the <code>push()</code> method adds elements to the end of an array, and <code>unshift()</code> adds elements to the beginning. Consider the following: 

```js
let twentyThree = 'XXIII';
let romanNumerals = ['XXI', 'XXII'];

romanNumerals.unshift('XIX', 'XX'); 
// now equals ['XIX', 'XX', 'XXI', 'XXII']
romanNumerals.push(twentyThree); 
// now equals ['XIX', 'XX', 'XXI', 'XXII', 'XXIII']
```

Notice that we can also pass variables, which allows us even greater flexibility in dynamically modifying our array's data.

## Instructions:
We have defined a function, <code>mixedNumbers</code>, which we are passing an array as an argument. Modify the function by using <code>push()</code> and <code>shift()</code> to add <code>'I', 2, 'three'</code>, to the beginning of the array and  <code>7, 'VIII', '9'</code> to the end so that the returned array contains representations of the numbers 1-9 in order.

## Challenge Seed:
```js
function mixedNumbers(arr) {
	// change code below this line

	// change code above this line
	return arr;
}

mixedNumbers(['IV', 5, 'six'])
```

## Challenge Solution
```js
function mixedNumbers(arr) {
	arr.push(7, 'VIII', '9');
	arr.unshift('I', 2, 'three');
	return arr;
}
```

## Tests:
```js
assert.deepEqual(mixedNumbers(['IV', 5, 'six']), ['I', 2, 'three', 'IV', 5, 'six', 7, 'VIII', '9'], "<code>mixedNumbers(['IV', 5, 'six'])</code> should now return <code>['I', 2, 'three', 'IV', 5, 'six', 7, 'VIII', '9']</code>");
assert.notStrictEqual(mixedNumbers.toString().search(/\.push\(/), -1, "The <code>mixedNumbers</code> function should utilize the <code>push()</code> method."); 
assert.notStrictEqual(mixedNumbers.toString().search(/\.unshift\(/), -1, "The <code>mixedNumbers</code> function should utilize the <code>unshift()</code> method."); 
```

**Comment by no-stack-dub-sack** [comment]
# Remove items from an array with  <code>pop()</code> and <code>shift()</code>

## Instructions
Both <code>push()</code> and <code>unshift()</code> have corresponding methods that are nearly functional opposites: <code>pop()</code> and <code>shift()</code>. As you may have guessed by now, instead of adding, <code>pop()</code> <em>removes</em> an element from the end of an array, while <code>shift()</code> removes an element from the beginning. The key difference between <code>pop()</code> and <code>shift()</code> and their cousins <code>push()</code> and <code>unshift()</code>, is that neither method takes parameters, and each only allows an array to be modified by a single element at a time.

Let's take a look: 

```js
let greetings = ['whats up?', 'hello', 'see ya!'];

greetings.pop() 
// now equals ['whats up?', 'hello']
greetings.shift() 
// now equals ['hello']
```

We can also return the value of the removed element with either method like this:

```js
let popped = greetings.pop(); 
// returns 'hello'
// greetings now equals []
```

## Instructions
We have defined a function, <code>popShift</code>, which takes an array as an argument and returns a new array. Modify the function, using <code>pop()</code> and <code>shift()</code>, to remove the first and last elements of the argument array, and assign the removed elements to their corresponding variables, so that the returned array contains their values.

## Challenge Seed
```js
function popShift(arr) {
    let popped = // change code here
    let shifted = // change code here
    return [shifted, popped];
}

popShift(['challenge', 'is', 'not', 'complete']);
```

## Challenge Solution
```js
function popShift(arr) {
    let popped = arr.pop();
    let shifted = arr.shift();
    return [shifted, popped];
}
```

## Tests
```js
assert.deepEqual(popShift(['challenge', 'is', 'not', 'complete']), ['challenge', 'complete'], "<code>popShift(['challenge', 'is', 'not', 'complete'])</code> should return <code>['challenge', 'complete']</code>.");
assert.notStrictEqual(popShift.toString().search(/\.pop\(/), -1, "The <code>popShift</code> function should utilize the <code>pop()</code> method."); 
assert.notStrictEqual(popShift.toString().search(/\.shift\(/), -1, "The <code>popShift</code> function should utilize the <code>shift()</code> method."); 
```

**Comment by no-stack-dub-sack** [comment]
## Remove items using <code>splice()</code>

### Instructions
Ok, so we've learned how to remove elements from the beginning and end of arrays using <code>pop()</code> and <code>shift()</code>, but what if we want to remove an element from somewhere in the middle? Or remove more than one element at once? Well, that's where <code>splice()</code> comes in. <code>splice()</code> allows us to do just that: <strong>remove any numnber of consecutive 
elements</strong> from anywhere on an array.

<code>splice()</code> can take up to 3 parameters, but for now, we'll focus on just the first 2. The first two parameters of <code>splice()</code> are integers which represent indexes, or postions, of the array that <code>splice()</code> is being called upon. And remember, arrays are <em>zero-indexed</em>, so to indicate the first element of an array, we would use <code>0</code>. <code>splice()</code>'s first parameter represents the index on the array from which to begin removing elements, while the second parameter indicates the number of elements to delete. For example: 
```js
let array = ['today', 'was', 'not', 'so', 'great'];
array.splice(2, 2); 
// indicates that splice() should remove 2 elements beginning with the 3rd element
// array now equals ['today', 'was', 'great']
```
<code>splice()</code> not only modifies the array it is being called on, but it also returns a new array containing the value of the removed elements. For example: 
```js
let array = ['today', 'was', 'not', 'so', 'great'];
let newArray = array.splice(3, 2); 
// newArray equals ['so', 'great']
```

### Instructions
We've defined a function, <code>sumOfTen</code>, which takes an array as an argument and returns the sum of that array's elements. Modify the function, using <code>splice()</code>, so that it returns a value of <code>10</code>.

### Challenge Seed
```js
function sumOfTen(arr) {
    // change code below this line

    // change code above this line
    return arr.reduce((a, b) => a + b);
}

sumOfTen([2, 5, 1, 5, 2, 1]);
```

### Challenge Solution
```js
function sumOfTen(arr) {
    arr.splice(1, 2);
    return arr.reduce((a, b) => a + b);
}
```

### Tests
```js
assert.strictEqual(sumOfTen([2, 5, 1, 5, 2, 1]), 10, "<code>sumOfTen</code> should return 10.");
assert.notStrictEqual(sumOfTen.toString().search(/\.splice\(/), -1, "The <code>sumOfTen</code> function should utilize the <code>splice()</code> method.")
```

**Comment by no-stack-dub-sack** [comment]
## Add items using <code>splice()</code>

### Instructions
Remember in the last challenge we mentioned that <code>splice()</code> can take up to three parameters? Well, we can go one step further with <code>splice()</code> &mdash; in addition to removing elements, we can use that third parameter, which represents one or more elements, to <em>add</em> them as well. This can be incredibly useful for quickly switching out an element, or a set of elements, for another. For instance, let's say you're storing a color scheme for a set of DOM elements in an array, and want to dynamically change a color based on some action:

```js
function colorChange(arr, index, newColor) {
	arr.splice(index, 1, newColor);
	return arr;
} 

let colorScheme = ['#878787', '#a08794', '#bb7e8c', '#c9b6be', '#d1becf'];
colorScheme = colorChange(colorScheme, 2, "#332327");
// we have removed '#bb7e8c' and added '#332327' in its place
// colorScheme now equals ['#878787', '#a08794', '#332327', '#c9b6be', '#d1becf']
```
This function takes an array of hex values, an index at which to remove an element, and the new color to replace the removed element with. The return value is an array containing a newly modified color scheme! While this example is a bit oversimplified, we can see the value that utilizing <code>splice()</code> to its maximum potential can have.

### Instructions
We have defined a function, <code>htmlColorNames</code>, which takes an array of html colors as an argument. Modify the function using <code>splice()</code> to remove the first two elements of the array and add <code>'DarkSalmon'</code> and <code>'BlanchedAlmond'</code> in their respective places.

### Challenge Seed
```js
function htmlColorNames(arr) {
    // change code below this line
    
    // change code above this line
    return arr;
} 

htmlColorNames(['DarkGoldenRod', 'WhiteSmoke', 'LavenderBlush', 'PaleTurqoise', 'FireBrick']);
```

### Challenge Solution
```js
function htmlColorNames(arr) {
    arr.splice(0, 2, 'DarkSalmon', 'BlanchedAlmond');
    return arr;
} 
```

### Tests
```js
assert.deepEqual(htmlColorNames(['DarkGoldenRod', 'WhiteSmoke', 'LavenderBlush', 'PaleTurqoise', 'FireBrick']), ['DarkSalmon', 'BlanchedAlmond', 'LavenderBlush', 'PaleTurqoise', 'FireBrick'], "<code>htmlColorNames</code> should return ['DarkSalmon', 'BlanchedAlmond', 'LavenderBlush', 'PaleTurqoise', 'FireBrick']");
assert.notStrictEqual(htmlColorNames.toString().search(/\.splice\(/), -1, "The <code>htmlColorNames</code> function should utilize the <code>splice()</code> method.")
```

**Comment by no-stack-dub-sack** [comment]
## Copy an array with <code>slice()</code>

### Instructions
The next method we will cover is <code>slice()</code>. <code>slice()</code>, rather than modifying an array, copies, or <em>extracts</em>, a given mumber of elements to a new array, leaving the array it is called upon untouched. <code>slice()</code> takes only 2 parameters &mdash; the first is the index at which to begin extraction, and the second is the index at which to stop extraction (extraction will occur up to, but not including the element at this index). Consider this:

```js
let weatherConditions = ['rain', 'snow', 'sleet', 'hail', 'clear'];
let todaysWeather = weatherConditions.slice(1, 3);
// todaysWeather equals ['snow', 'sleet'];
// weatherConditions still equals ['rain', 'snow', 'sleet', 'hail', 'clear']
```
In effect, we have created a new array by extracting elements from an existing array.

### Instructions
Instructions:</strong> We have defined a function, <code>forecast</code>, that takes an array as an argument. Modify the function using <code>slice()</code> to extract information from the argument array and return a new array that contains the elements <code>'warm'</code> and <code>'sunny'</code>.

### Challenge Seed
```js
function forecast(arr) {
    // change code below this line
	
    return arr;
}

// do not change code below this line
forecast(['cold', 'rainy', 'warm', 'sunny', 'cool', 'thunderstorms']);
```

### Challenge Solution
```js
function forecast(arr) {
    return arr.slice(2, 4);
}
```

### Tests
```js
assert.deepEqual(forecast(['cold', 'rainy', 'warm', 'sunny', 'cool', 'thunderstorms']), ['warm', 'sunny'], "<code>forecast</code> should return <code>['warm', 'sunny'].");
assert.notStrictEqual(forecast.toString().search(/\.slice\(/), -1, "The <code>forecast</code> function should utilize the <code>slice()</code> method.");
```

**Comment by no-stack-dub-sack** [comment]
## Copy an array with spread syntax

### Instructions
While <code>slice()</code> allows us to be selective about what elements of an array to copy, ammong several other useful tasks, ES6's new spread syntax allows us to easily copy <em>all</em> of an array's elements, in order, with a simple and highly readable syntax. The spread syntax simply looks like this: <code>...</code>

In practice, we can use the spread syntax to copy an array like so:

```js
let thisArray = [true, true, undefined, false, null];
let thatArray = [...thisArray];
// thatArray equals [true, true, undefined, false, null]
// thisArray remains unchanged, and is identical to thatArray
```

### Instructions
We have defined a function, <code>copyMachine</code> which takes <code>arr</code> (an array) and <code>num</code> (a number) as arguments. The function is supposed to return a new array made up of <code>num</code> copies of <code>arr</code>. We have done most of the work for you, but it doesn't work quite right yet. Modidy the function using the spead syntax so that it works correctly (hint: another method we have already covered might come in handy here!).

### Challenge Seed
```js
function copyMachine(arr, num) {
    let newArr = [];
    while (num >= 1) {
	// change code below this line

	// change code above this line
	num--;
    }
    return newArr;
}

copyMachine([true, false, true], 2);
```

### Challenge Solution
```js
function copyMachine(arr, num) {
    let newArr = [];
    while (num >= 1) {
	newArr.push([...arr]);
	num--;
    }
    return newArr;
}
```

### Tests
```js
assert.deepEqual(copyMachine([true, false, true], 2), [[true, false, true], [true, false, true]], "<code>copyMachine([true, false, true], 2)</code> should return <code>[[true, false, true], [true, false, true]]</code>.")
assert.deepEqual(copyMachine([1, 2, 3], 5), [[1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3]], "<code>copyMachine([1, 2, 3], 5)</code> should return <code>[[1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3]]</code>.");
assert.deepEqual(copyMachine([true, true, null], 1), [[true, true, null]], "<code>copyMachine([true, true, null], 1)</code> should return <code>[[true, true, null]]</code>.");
assert.deepEqual(copyMachine(['it works'], 3), [['it works'], ['it works'], ['it works']], "<code>copyMachine(['it works'], 3)</code> should return <code>[['it works'], ['it works'], ['it works']]</code>.");
asser
```

**Comment by no-stack-dub-sack** [comment]
## Check for the presence of an element with <code>indexOf()</code>

### Instructions
Since arrays can be changed, or <em>mutated</em>, at any time, there's no guarantee about where a particular piece of data will be on a given array, or if that element even still exists. Luckily, JavaScript provides us with another built-in method, <code>indexOf()</code>, that allows us to quickly and easily check for the presence of an element on an array. <code>indexOf()</code> takes an element as a parameter, and when called, it returns the position, or index, of that element, or <code>-1</code> if the element does not exist on the array. 

For example:

```js
let fruits = ['apples', 'pears', 'oranges', 'peaches', 'pears'];
fruits.indexOf('dates') // returns -1
fruits.indexOf('oranges') // returns 2
fruits.indexOf('pears') // returns 1, the first index at which the element exists
```

### Instructions
<code>indexOf()</code> can be incredibly useful for quickly checking for the presence of an element on an array. We have defined a function, <code>quickCheck</code>, that takes an array and an element as arguments. Modify the function using <code>indexOf()</code> so that it returns true if the passed element exists on the array, and false if it does not.

### Challenge Seed
```js
function quickCheck(arr, elem) {
    // change code below this line

    return elem;
}

quickCheck(['squash', 'onions', 'shallots'], 'mushrooms');
```

### Challenge Solution
```js
function quickCheck(arr, elem) {
    if (arr.indexOf(elem) !== -1) {
	return true
    } else {
	return false;
    }
}
```

### Tests
```js
assert.strictEqual(quickCheck(['squash', 'onions', 'shallots'], 'mushrooms'), false, "<code>quickCheck(['squash', 'onions', 'shallots'], 'mushrooms')</code> should return <code>false</code>.");
	assert.strictEqual(quickCheck(['squash', 'onions', 'shallots'], 'onions'), true, "<code>quickCheck(['squash', 'onions', 'shallots'], 'onions')</code> should return <code>true</code>.");
	assert.strictEqual(quickCheck([3, 5, 9, 125, 45, 2], 125), true, "<code>quickCheck([3, 5, 9, 125, 45, 2], 125)</code> should return <code>true</code>.");
	assert.strictEqual(quickCheck([true, false, false], undefined), false, "<code>quickCheck([true, false, false], undefined)</code> should return <code>false</code>.");
	assert.notStrictEqual(quickCheck.toString().search(/\.indexOf\(/), -1, "The <code>quickCheck</code> function should utilize the <code>indexOf() method.</code>");
```

**Comment by no-stack-dub-sack** [comment]
## Iterate through all the items in an array

### Instructions
Sometimes when working with arrays, it is very handy to be able to iterate through each item to find one or more elements that we might need, or to manipulate an array based on which data items meet a certain set of requirements. JavaScript offers several built in methods that each iterate over arrays in slightly different ways to achieve different results (such as <code>every()</code>, <code>forEach()</code>, <code>map()</code>, etc.), however the technique which is most flexible and offers us the greatest amount of control is a simple <code>for</code> loop.<br><br>

For example:

```js
function greaterThanTen(arr) {
    let newArr = [];
    for (let i = 0; i < arr.length; i++) {
	if (arr[i] > 10) {
		newArr.push(arr[i]);
	}
    }
    return newArr
}

greaterThanTen([2, 12, 8, 14, 80, 0, 1]);
// returns [12, 14, 80]
```
Using a <code>for</code> loop, this function iterates through and accesses each element of the array, and subjects it to a simple test that we have created. In this way, we have easily and programmatically determined which data items are greater than 10, and returned a new array containing those items.

### Instructions
We have defined a function, <code>filteredArray</code>, which takes <code>arr</code>, a nested array, and <code>elem</code> as arguments, and returns a new array. <code>elem</code> represents an element that may or may not be present on one or more of the arrays nested within <code>arr</code>. Modify the function, using a <code>for</code> loop, to return a filtered version of the passed array such that any array nested within <code>arr</code> containing <code>elem</code> has been removed.

### Challenge Seed
```js
function filteredArray(arr, num) {
    newArr = [];
    // change code below this line

    // change code above this line
    return newArr;
}

filteredArray([[3, 2, 3], [1, 6, 3], [3, 13, 26], [19, 3, 9]], 3);
```

### Challenge Solution
```js
function filteredArray(arr, num) {
    let newArr = [];
    for (let i = 0; i < arr.length; i++) {
	if (arr[i].indexOf(num) === -1) {
      	    newArr.push(arr[i]);
	}
    }
    return newArr;
}
```

### Tests
```js
assert.deepEqual(filteredArray([ [10, 8, 3], [14, 6, 23], [3, 18, 6] ], 18), [[10, 8, 3], [14, 6, 23]], "<code>filteredArray([[10, 8, 3], [14, 6, 23], [3, 18, 6]], 18)</code> should return <code>[[10, 8, 3], [14, 6, 23]]</code>");
assert.deepEqual(filteredArray([ ['trumpets', 2], ['flutes', 4], ['saxaphones', 2] ], 2), [['flutes', 4]], "<code>filteredArray([['trumpets', 2], ['flutes', 4], ['saxaphones'], 2], 2)</code> should return <code>[['flutes', 4]]</code>");
assert.deepEqual(filteredArray([['amy', 'beth', 'sam'], ['dave', 'sean', 'peter']], 'peter'), [['amy', 'beth', 'sam']], "<code>filteredArray([['amy', 'beth', 'sam'], ['dave', 'sean', 'peter']], 'peter')</code> should return [['amy', 'beth', 'sam']].");
assert.deepEqual(filteredArray([[3, 2, 3], [1, 6, 3], [3, 13, 26], [19, 3, 9]], 3), [], "<code>filteredArray([[3, 2, 3], [1, 6, 3], [3, 13, 26], [19, 3, 9]], 3)</code> should return <code>[]</code>.");
assert.notStrictEqual(filteredArray.toString().search(/for/), -1, "The <code>filteredArray</code> function should utilize a <code>for</code> loop.");
```

**Comment by no-stack-dub-sack** [comment]
## Create complex nested arrays 

### Instructions
One of the most powerful features when thinking of arrays as data structures, is that arrays can contain, or even be completely made up of other arrays. We have seen arrays that contain arrays in previous challenges, but fairly simple ones. However, arrays can contain an infinite depth of arrays that can contain other arrays, each with their own arbitrary levels of depth, and so on. In this way, an array can very quickly become very complex data structure,  Consider the following example:
```js
let nestedArray = [ // top, or first level - the outer most array
	['deep'], // an array within an array, 2 levels of depth
	[ 
		['deeper'], ['deeper'] // 2 arrays nested 3 levels deep
	], 
	[ 
		[ 
			['deepest'], ['deepest'] // 2 arrays nested 4 levels deep 
		], 
		[ 
			[
				['deepest-est?'] // an array nested 5 levels deep
			] 
		]
	], 
];
```
While this example may seem convoluted, this level of complexity is not unheard of, or even unusual, when dealing with large amounts of data. 

However, we can still very easily access the deepest levels of an array this complex with bracket notation:
```js
console.log(nestedArray[2][1][0][0][0]);
logs 'deepest-est?'
```
And now that we know where that piece of data is, we can reset it if we need too:
```js
nestedArray[2][1][0][0][0] = 'deeper still';
console.log(nestedArray[2][1][0][0][0]);
now logs 'deeper still'
```
### Instructions
We have defined a variable, <code>myNestedArray</code>, set equal to an array. Modify <code>myNestedArray</code>, using any combination of strings, numbers, and booleans for data elements, so that it has exactly three levels of depth (remember, the outer-most array is level 1).

### Challenge Seed
```js
let myNestedArray = [
	
	// change code here
	
];
```

### Challenge Solution
```js
let myNestedArray = [
	[
		1, 2, 3, 3, 5, 6, [true, false, false, 12]
	]
];
```

### Tests
```js
assert.strictEqual((function(arr) { let flattened = (function flatten(arr) { const flat = [].concat(...arr); return flat.some (Array.isArray) ? flatten(flat) : flat; })(arr); for (let i = 0; i < flattened.length; i++) { if ( typeof flattened[i] !== 'number' && typeof flattened[i] !== 'string' && typeof flattened[i] !== 'boolean') { return false } } return true })(myNestedArray), true, "<code>myNestedArray</code> should contain only numbers, booleans, and strings as data elements.");
assert((function(arr) { for (let i = 0; i < arr.length; i++) { if (Array.isArray(arr[i])) {for (let j = 0; j < arr[i].length; j++) { if (Array.isArray(arr[i][j])) { return true } } } } })(myNestedArray) === true && (function(arr) { for (let i = 0; i < arr.length; i++) { for (let j = 0; j < arr[i].length; j++) { for (let k = 0; k < arr[i][j].length; k++) { if ( Array.isArray(arr[i][j][k]) ) { return false } } } } })(myNestedArray) === undefined, "<code>myNestedArray</code> should have exactly 3 levels of depth.");
```

**Comment by no-stack-dub-sack** [comment]
## Combining arrays with spread syntax

### Instructions
Another huge advantage of the spread syntax, is the ability to combine arrays, or to insert all the elements of one array into another, at any index. With more traditional syntaxes, we can concatenate arrays, but this only allows us to combine arrays at the end of one, and at the start of another. Spread syntax makes the following operation extremely simple:

```js
let thisArray = ['sage', 'rosemary', 'parsely', 'thyme'];

let thatArray = ['basil', 'cilantro', ...thisArray, 'corriander'];
// thatArray now equals ['basil', 'cilantro', 'sage', 'rosemary', 'parsely', 'thyme', 'corriander']
```
Using spread syntax, we have just achieved an operation that would have been more more complex and more verbose had we used traditional methods.

### Instructions
We have defined a function <code>spreadOut</code> that returns the variable <code>sentence</code>, modify the function using the spread syntax so that it returns the array <code>['learning', 'to', 'code', 'is', 'fun']</code>.

### Challenge Seed
```js
function spreadOut() {
    let fragment = ['to', 'code'];
    let sentence = 'change code here'
    return sentence;
}

spreadOut();
```

### Challenge Solution
```js
function spreadOut() {
    let fragment = ['to', 'code'];
    let sentence = ['learning', ...fragment, 'is', 'fun'];
    return sentence;
}
```

### Tests
```js
assert.deepEqual(spreadOut(), ['learning', 'to', 'code', 'is', 'fun'], "<code>spreadOut</code> should return <code>['learning', 'to', 'code', 'is', 'fun']</code>.");
assert.notStrictEqual(spreadOut.toString().search(/\.\.\./), -1, "The <code>spreadOut</code> function should utilize the spread syntax.");
```

**Comment by no-stack-dub-sack** [comment]
@QuincyLarson, @bonham000 Ok! These array challenges are now done, I've just posted mine and all challenges are linked to their titles in the OP. Really looking forward to checking all this stuff out in beta!

**Comment by QuincyLarson** [comment]
@no-stack-dub-sack awesome work! I've been reading through these, and they look great. I'll start importing them into the seed file.

@smileytechguy I'm going to go ahead and close this issue and if you want to create a "Insert elements with spread syntax" later, we can add this later.

**Comment by no-stack-dub-sack** [comment]
@QuincyLarson I went ahead and added that challenge because I wasn't sure if we were waiting on this, I'm sure there's improvements that can be made though if @smileytechguy wants to take a look and help QA these!

**Comment by QuincyLarson** [comment]
@no-stack-dub-sack awesome! I didn't see that. OK - we're good to go then. 

@smileytechguy there will be tons of opportunities to improve and expand our curriculum once we get these live.

**Comment by smileytechguy** [comment]
These look pretty good!

**Comment by LudwigJedde** [comment]
> ## Add items to an array with `push()` and `unshift()`
> ### Description:
> An array's length, like the data types it can contain, is not fixed. Arrays can be defined with a length of any number of elements, and elements can be added or removed over time. In this challenge, we will look at two methods with which we can programmatically modify an array: `Array.push()` and `Array.unshift()`. 
> 
> 
> Both methods take one or more elements as _parameters_ and add those elements to the array the method is being called on; the `push()` method adds elements to the end of an array, and `unshift()` adds elements to the beginning. Consider the following:
> 
> ```js
> let twentyThree = 'XXIII';
> let romanNumerals = ['XXI', 'XXII'];
> 
> romanNumerals.unshift('XIX', 'XX'); 
> // now equals ['XIX', 'XX', 'XXI', 'XXII']
> romanNumerals.push(twentyThree); 
> // now equals ['XIX', 'XX', 'XXI', 'XXII', 'XXIII']
> ```
> 
> Notice that we can also pass variables, which allows us even greater flexibility in dynamically modifying our array's data.
> 
> ### Instructions:
> We have defined a function, `mixedNumbers`, which we are passing an array as an argument. Modify the function by using `push()` and `shift()` to add `'I', 2, 'three'`, to the beginning of the array and `7, 'VIII', '9'` to the end so that the returned array contains representations of the numbers 1-9 in order.
> 
> ### Challenge Seed:
> ```js
> function mixedNumbers(arr) {
>     // change code below this line
> 
>     // change code above this line
>     return arr;
> }
> 
> // do not change code below this line
> mixedNumbers(['IV', 5, 'six'])
> ```
> 
> ### Challenge Solution
> ```js
> function mixedNumbers(arr) {
>     arr.push(7, 'VIII', '9');
>     arr.unshift('I', 2, 'three');
>     return arr;
> }
> ```
> 
> ### Tests:
> ```js
> assert.deepEqual(mixedNumbers(['IV', 5, 'six']), ['I', 2, 'three', 'IV', 5, 'six', 7, 'VIII', '9'], "<code>mixedNumbers(['IV', 5, 'six'])</code> should now return <code>['I', 2, 'three', 'IV', 5, 'six', 7, 'VIII', '9']</code>");
> assert.notStrictEqual(mixedNumbers.toString().search(/\.push\(/), -1, "The <code>mixedNumbers</code> function should utilize the <code>push()</code> method."); 
> assert.notStrictEqual(mixedNumbers.toString().search(/\.unshift\(/), -1, "The <code>mixedNumbers</code> function should utilize the <code>unshift()</code> method."); 
> ```

**you have to put this on to validate the exercise**
9 and not '9' in arr.push 

function mixedNumbers(arr) {
>     arr.push(7, 'VIII', 9 );
>     arr.unshift('I', 2, 'three');
>     return arr;

**Comment by QuincyLarson** [comment]
@LudwigJedde Thanks for catching this. Could you create a GitHub issue on github.com/freecodecamp/freecodecamp instead? This repo is archived. Then we can tackle fixing this.
